language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -135,16 +135,18 @@ computeCek con (Var _ varName)          = do
 -- 3. puts 'FrameApplyFun' on top of the context and proceeds with the argument from 'FrameApplyArg'
 -- 4. grows the resulting term ('FrameWrap')
 returnCek :: Context -> Plain Value -> CekM EvaluationResultDef
-returnCek []                                  res = pure $ EvaluationSuccess res
-returnCek (FrameTyInstArg ty           : con) fun = instantiateEvaluate con ty fun
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
-    funVarEnv <- getVarEnv
-    withVarEnv argVarEnv $ computeCek (FrameApplyFun funVarEnv fun : con) arg
-returnCek (FrameApplyFun funVarEnv fun : con) arg = do
+returnCek []                                  Error{}     = pure EvaluationFailure
+returnCek []                                  res         = pure $ EvaluationSuccess res
+returnCek (FrameApplyFun funVarEnv fun : con) arg         = do
     argVarEnv <- getVarEnv
     applyEvaluate funVarEnv argVarEnv con fun arg
-returnCek (FrameIWrap ann pat arg      : con) val = returnCek con $ IWrap ann pat arg val
-returnCek (FrameUnwrap                 : con) dat = case dat of
+returnCek (_ : con)                           err@Error{} = returnCek con err
```

Comment:
So the idea is that we do not throw an `Error` immediately, but give the `FrameApplyFun` frame a chance to handle the error. For any other frame we let the error pop up, but we do not change the type that `Error` carries, so this breaks subjection reduction (don't tell Phil I've written something like that!). If the evaluator was untyped, I believe it would make sense to adapt the CEK machine this way, but breaking subject reduction is a too terrible of a sin.

Reply:
Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions, e.g. https://github.com/input-output-hk/plutus/issues/1145#issuecomment-503954247. So I'm not sure how much of a sin this is.

Reply:
Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions, e.g. https://github.com/input-output-hk/plutus/issues/1145#issuecomment-503954247. So I'm not sure how much of a sin this is.

Reply:
> Note that we already don't return terms whose internal types are correct because we don't actually do type substitutions

Yeah, but I think it's also a bug. But that bug is easily fixable unlike the one that I introduced here.

Reply:
Hmmm, I'm not sure. See also https://github.com/input-output-hk/plutus/issues/948